tests/integration/rosetta/rosetta_test.go

```diff
@@ -0,0 +1,272 @@
+// +build integration
+
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta_test
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/labstack/echo/v4"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/api/rosetta"
+	"github.com/optakt/flow-dps/models/identifier"
+	"github.com/optakt/flow-dps/service/state"
+
+	"github.com/optakt/flow-dps/rosetta/invoker"
+	"github.com/optakt/flow-dps/rosetta/retriever"
+	"github.com/optakt/flow-dps/rosetta/scripts"
+	"github.com/optakt/flow-dps/rosetta/validator"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+const testChainID = "flow-testnet"
+
+// initialize the rosetta Data HTTP handler once and reuse for all tests.
+var rosettaSvc *rosetta.Data
+
+func setupDB() (*rosetta.Data, *badger.DB, error) {
+
+	const testDbBackupFile = "rosetta-get-balance-test.db"
+	opts := badger.DefaultOptions("").
+		WithInMemory(true).
+		WithReadOnly(true).
+		WithLogger(nil)
+
+	db, err := badger.Open(opts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dbSnapshot := hex.NewDecoder(strings.NewReader(getDBSnapshot()))
+
+	err = db.Load(dbSnapshot, 10)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	core, err := state.NewCoreFromDB(db)
+	if err != nil {
+		return nil, nil, err
+
+	}
+
+	// setup scaffolding for minimal server we need for the tests
+	params, ok := dps.FlowParams[flow.ChainID(testChainID)]
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid chain")
+	}
+
+	generator := scripts.NewGenerator(params)
+	invoke := invoker.New(zerolog.Nop(), core)
+	validate := validator.New(params, core.Height())
+	retrieve := retriever.New(generator, invoke)
+
+	svc := rosetta.NewData(validate, retrieve)
+
+	return svc, db, err
+}
+
+func TestMain(m *testing.M) {
+
+	svc, db, err := setupDB()
+	if err != nil {
+		log.Fatalf("could not perform setup for test: %v", err)
+	}
+
+	rosettaSvc = svc
+
+	defer db.Close()
+
+	m.Run()
```

Comment:
I just realized yesterday, it should be `os.Exit(m.Run())` because `m.Run()` returns an `int` which will be `0` if tests were successful and `1` otherwise. If we don't check the return value of `m.Run()`, running `go test` will show an error but the return value won't be correct (assuming we only run this specific test)

Reply:
That was in my first version, I removed it when I refactored it to move work to `setupDB`. I removed it once again because of defers.

Even though the docs do use `os.Exit(m.Run())` , from what I saw, exit code get set to 0 on success or 1 on error either way (tested on my machine).

Checking the docs:

>  TestMain runs in the main goroutine and can do whatever setup and teardown is necessary around a call to m.Run. m.Run will return an exit code that may be passed to os.Exit. **If TestMain returns, the test wrapper will pass the result of m.Run to os.Exit itself.**

The last part sounds like test wrapper will set the exit code automagically.. Not sure about the "if test main returns" qualifier though, does that mean on crashes?

Reply:
[It is not very clear indeed](https://golang.org/src/testing/testing.go?s=45401:45429#L1332)

From the code, it seems like that might be the case, but I'm not sure exactly where it is that our call to `m.Run` gets injected.

Still, since they write `os.Exit(m.Run())` on their side and since it's done like this in most places I've found, I think we should do the same.